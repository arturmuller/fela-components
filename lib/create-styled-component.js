import React from "react";
import R from "ramda";
import T from "prop-types";

export default function createStyledComponent(displayName, use) {
  return class StyledComponent extends React.Component {
    static displayName = displayName;

    static propTypes = {
      use: T.oneOfType([T.string, T.func]),
      visual: T.oneOfType([T.object, T.func]),
      animations: T.objectOf(T.object),
      innerRef: T.func,
      children: T.node,
    };

    static defaultProps = {
      use,
      visual: {},
    };

    static contextTypes = {
      renderer: T.object.isRequired,
      theme: T.object,
    };

    renderClassName = (visual, animations) => {
      const { renderer, theme = {} } = this.context;
      if (typeof visual === "function") {
        return renderer.renderRule(visual, {
          animations: this.renderKeyframes(animations),
          theme,
        });
      }

      return renderer.renderRule(() => visual);
    };

    renderKeyframes = (animations = {}) => {
      const { renderer } = this.context;
      return R.map(
        animation => renderer.renderKeyframe(() => animation),
        animations
      );
    };

    render() {
      const { use: Component, visual, animations, innerRef, ...props } = this.props;
      const ref = innerRef ? { ref: r => innerRef(r) } : {}
      return (
        <Component
          {...ref}
          {...props}
          className={this.renderClassName(visual, animations)}
        />
      );
    }
  };
}
