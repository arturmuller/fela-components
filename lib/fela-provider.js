import React from "react";
import FelaDOM from "fela-dom";
import T from "prop-types";

export default class FelaProvider extends React.Component {
  static propTypes = {
    renderer: T.object.isRequired,
    mountNode: T.any,
    staticStyles: T.string,
    fonts: T.arrayOf(
      T.shape({
        family: T.string.isRequired,
        files: T.arrayOf(T.string.isRequired).isRequired,
        properties: T.object,
      })
    ),
    theme: T.object,
    children: T.element,
  };

  static defaultProps = {
    theme: {},
  };

  static childContextTypes = {
    renderer: T.object,
    theme: T.object,
  };

  getChildContext() {
    const { renderer, theme } = this.props;
    return { renderer, theme };
  }

  componentDidMount() {
    const { staticStyles, fonts, mountNode, renderer } = this.props;

    if (staticStyles) {
      renderer.renderStatic(staticStyles);
    }

    if (fonts) {
      fonts.forEach(({ family, files, ...properties }) =>
        renderer.renderFont(family, files, properties)
      );
    }

    if (mountNode) {
      FelaDOM.render(renderer, mountNode);
    }
  }

  render() {
    return React.Children.only(this.props.children);
  }
}
