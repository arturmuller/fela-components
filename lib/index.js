export default from "./styled-components";
export * from "./styled-components";
export createStyledComponent from "./create-styled-component";
export FelaProvider from "./fela-provider";
export ThemeProvider from "./theme-provider";
