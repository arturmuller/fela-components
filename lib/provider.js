import React from "react";
import { render } from "fela-dom";
import T from "prop-types";

class Provider extends React.Component {
  static propTypes = {
    renderer: T.object.isRequired,
    staticStyles: T.string,
    fonts: T.arrayOf(
      T.shape({
        family: T.string.isRequired,
        files: T.arrayOf(T.string.isRequired).isRequired,
        properties: T.object,
      })
    ),
    children: T.element,
  };

  static childContextTypes = {
    renderer: T.object,
  };

  getChildContext() {
    const { renderer } = this.props;
    return { renderer };
  }

  componentDidMount() {
    const { staticStyles, fonts, renderer } = this.props;

    if (staticStyles) {
      renderer.renderStatic(staticStyles);
    }

    if (fonts) {
      fonts.forEach(({ family, files, ...properties }) =>
        renderer.renderFont(family, files, properties)
      );
    }

    render(renderer);
  }

  render() {
    return React.Children.only(this.props.children);
  }
}

// FelaProvider is legacy export
export { Provider, Provider as FelaProvider };
