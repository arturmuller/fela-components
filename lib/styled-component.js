import React from "react";
import R from "ramda";
import T from "prop-types";

export default class StyledComponent extends React.Component {
  static propTypes = {
    use: T.oneOfType([T.string, T.func]),
    visual: T.oneOfType([T.object, T.func]),
    animations: T.objectOf(T.object),
    children: T.node,
  };

  static defaultProps = {
    use: "div",
    visual: {},
  };

  static contextTypes = {
    renderer: T.object.isRequired,
    theme: T.object,
  };

  renderClassName = (visual, animations) => {
    const { renderer, theme = {} } = this.context;
    if (typeof visual === "function") {
      return renderer.renderRule(visual, {
        animations: this.renderKeyframes(animations),
        theme,
      });
    }

    return renderer.renderRule(() => visual);
  };

  renderKeyframes = (animations = {}) => {
    const { renderer } = this.context;
    return R.map(
      animation => renderer.renderKeyframe(() => animation),
      animations
    );
  };

  render() {
    const { use: Component, visual, animations, ...props } = this.props;
    return (
      <Component
        {...props}
        className={this.renderClassName(visual, animations)}
      />
    );
  }
}
