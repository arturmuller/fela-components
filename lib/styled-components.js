import createStyledComponent from "./create-styled-component";

// Generic
export const StyledComponent = createStyledComponent("StyledComponent", "div");
export default StyledComponent; // Legacy export

// Inline Elements
export const Anchor = createStyledComponent("Anchor", "a");
export const Button = createStyledComponent("Button", "button");
export const Emphasis = createStyledComponent("Emphasis", "em");
export const Image = createStyledComponent("Image", "img");
export const Input = createStyledComponent("Input", "input");
export const Label = createStyledComponent("Label", "label");
export const Select = createStyledComponent("Select", "select");
export const Span = createStyledComponent("Span", "span");
export const Strong = createStyledComponent("Strong", "strong");
export const Textarea = createStyledComponent("Textarea", "textarea");

// Block Elements
export const Aside = createStyledComponent("Aside", "aside");
export const Blockquote = createStyledComponent("Blockquote", "blockquote");
export const Division = createStyledComponent("Division", "div");
export const Footer = createStyledComponent("Footer", "footer");
export const Form = createStyledComponent("Form", "form");
export const Header = createStyledComponent("Header", "header");
export const HeadingFive = createStyledComponent("HeadingFive", "h5");
export const HeadingFour = createStyledComponent("HeadingFour", "h4");
export const HeadingOne = createStyledComponent("HeadingOne", "h1");
export const HeadingSix = createStyledComponent("HeadingSix", "h6");
export const HeadingThree = createStyledComponent("HeadingThree", "h3");
export const HeadingTwo = createStyledComponent("HeadingTwo", "h2");
export const Main = createStyledComponent("Main", "main");
export const Nav = createStyledComponent("Nav", "nav");
export const OrderedList = createStyledComponent("OrderedList", "ol");
export const Paragraph = createStyledComponent("Paragraph", "p");
export const UnorderedList = createStyledComponent("UnorderedList", "ul");
