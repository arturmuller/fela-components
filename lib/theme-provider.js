import React from "react";
import T from "prop-types";

export default class ThemeProvider extends React.Component {
  static propTypes = {
    theme: T.object,
    children: T.element,
  };

  static contextTypes = {
    theme: T.object,
  };

  static childContextTypes = {
    theme: T.object,
  };

  getChildContext() {
    const { context, props } = this;
    return { theme: { ...context.theme, ...props.theme } };
  }

  render() {
    return React.Children.only(this.props.children);
  }
}
