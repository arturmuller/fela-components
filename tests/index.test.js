import React from "react";
import { felaSnapshot } from "./utils";

import { Anchor } from "../lib/index";

it("Renders styles correctly", () => {
  const snapshot = felaSnapshot(
    <Anchor
      visual={{ display: "flex", alignItems: "center" }}
      href="http://www.facebook.com">
      Facebook
    </Anchor>
  );

  expect(snapshot).toMatchSnapshot();
});

it("Renders animations correctly", () => {
  const snapshot = felaSnapshot(
    <Anchor
      animations={{ fadeIn: { from: { opacity: 0 }, to: { opacity: 1 } } }}
      visual={({ animations }) => ({
        display: "flex",
        alignItems: "center",
        animationName: animations.fadeIn,
        "@media (min-width: 300px)": {
          color: "white",
          "@media (min-width: 400px)": {
            color: "black",
          },
        },
      })}
      href="http://www.facebook.com">
      Facebook
    </Anchor>
  );

  expect(snapshot).toMatchSnapshot();
});
