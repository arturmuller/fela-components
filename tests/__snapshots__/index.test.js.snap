// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders animations correctly 1`] = `
<a
  className="a b c d e"
  href="http://www.facebook.com"
>
  Facebook
</a>

@-webkit-keyframes k1 {
  from { opacity: 0; }
  to { opacity: 1; }
}
@-moz-keyframes k1 {
  from { opacity: 0; }
  to { opacity: 1; }
}
@keyframes k1 {
  from { opacity: 0; }
  to { opacity: 1; }
}
.a { display: flex; }
.b { align-items: center; }
.c { animation-name: k1; }
@media (min-width: 300px) {
  .d { color: white; }
}
@media (min-width: 300px) and (min-width: 400px) {
  .e { color: black; }
}

`;

exports[`Renders styles correctly 1`] = `
<a
  className="a b"
  href="http://www.facebook.com"
>
  Facebook
</a>

.a { display: flex; }
.b { align-items: center; }

`;
